driver/generic/st77xx.py

```diff
@@ -0,0 +1,417 @@
+ST77XX_MADCTL_MX = const(0x40)  # column address order (0: left to right; 1: right to left)
+ST77XX_MADCTL_MV = const(0x20)  # page/column order (0: normal mode 1; reverse mode)
+ST77XX_MADCTL_ML = const(0x10)  # line address order (0: refresh to to bottom; 1: refresh bottom to top)
+ST77XX_MADCTL_BGR = const(0x08) # colors are BGR (not RGB)
+ST77XX_MADCTL_RTL = const(0x04) # refresh right to left
+ST77XX_MADCTL_ROTS=(
+    const(0x00),                                # 0 = portrait
+    const(ST77XX_MADCTL_MX | ST77XX_MADCTL_MV), # 1 = landscape
+    const(ST77XX_MADCTL_MY | ST77XX_MADCTL_MX), # 2 = inverted portrait
+    const(ST77XX_MADCTL_MY | ST77XX_MADCTL_MV), # 3 = inverted landscape
+)
+
+ST77XX_COLOR_MODE_65K = const(0x50)
+ST77XX_COLOR_MODE_262K = const(0x60)
+ST77XX_COLOR_MODE_12BIT = const(0x03)
+ST77XX_COLOR_MODE_16BIT = const(0x05)
+ST77XX_COLOR_MODE_18BIT = const(0x06)
+ST77XX_COLOR_MODE_16M = const(0x07)
+
+
+ST77XX_COL_ROW_MODEL_START_ROTMAP={
+    # ST7789
+    (240,320,None):[(0,0),(0,0),(0,0),(0,0)],
+    (240,240,None):[(0,0),(0,0),(0,80),(80,0)],
+    (135,240,None):[(52,40),(40,53),(53,40),(40,52)],
+    # ST7735
+    (128,160,'blacktab'):[(0,0),(0,0),(0,0),(0,0)],
+    (128,160,'redtab'):[(2,1),(1,2),(2,1),(1,2)],
+}
+
+ST77XX_PORTRAIT = const(0)
+ST77XX_LANDSCAPE = const(1)
+ST77XX_INV_PORTRAIT = const(2)
+ST77XX_INV_LANDSCAPE = const(3)
+
+class St77xx_hw(object):
+    def __init__(self, *, cs, dc, spi, res, suppRes, bl=None, model=None, suppModel=[], rst=None, rot=ST77XX_LANDSCAPE, bgr=False, rp2_dma=None):
+        '''
+        This is an abstract low-level driver the ST77xx controllers, not to be instantiated directly.
+        Derived classes implement chip-specific bits. THe following parameters are recognized:
+
+        * *cs*: chip select pin (= slave select, SS)
+        * *dc*: data/command pin
+        * *bl*: backlight PWM pin (optional)
+        * *model*: display model, to account for variations in products
+        * *rst*: optional reset pin
+        * *res*: resolution tuple; (width,height) with zero rotation
+        * *rot*: display orientation (0: portrait, 1: landscape, 2: inverted protrait, 3: inverted landscape); the constants ST77XX_PORTRAIT, ST77XX_LANDSCAPE, ST77XX_INV_POTRAIT, ST77XX_INV_LANDSCAPE may be used.
+        * *bgr*: color order if BGR (not RGB)
+        * *rp2_dma*: optional DMA object for the rp2 port
+
+
+        Subclass constructors (implementing concrete chip) set in addition the following, not to be used directly:
+
+        * *suppModel*: models supported by the hardware driver
+        * *suppRes*: resolutions supported by the hardware driver, as list of (width,height) tuples
+        '''
+        self.buf1 = bytearray(1)
+        self.buf2 = bytearray(2)
+        self.buf4 = bytearray(4)
+
+        self.cs,self.dc,self.rst=[(machine.Pin(p,machine.Pin.OUT) if isinstance(p,int) else p) for p in (cs,dc,rst)]
+        if isinstance(self.bl,int): self.bl=machine.PWM(self.bl)
```

Comment:
I believe that in most Micropython ports PWM expects a `machine.Pin` object and not an `int`.  
See [ESP32 PWM for example](https://docs.micropython.org/en/latest/esp32/tutorial/pwm.html#pulse-width-modulation).

I'm not sure why it works with an `int` on rp2 (did you test this?)

Reply:
`machine.PWM(int)` [is possibly prot-specific](https://docs.micropython.org/en/latest/library/machine.PWM.html#machine.PWM) so I handle that manually now.